archunit/src/main/java/com/tngtech/archunit/lang/conditions/JoinCondition.java

```diff
@@ -0,0 +1,139 @@
+
+    List<ConditionWithEvents<T>> evaluateConditions(T item) {
+        return conditions.stream().map(condition -> new ConditionWithEvents<>(condition, item)).collect(toList());
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + "{" + conditions + "}";
+    }
+
+    static class ConditionWithEvents<T> {
+        private final ArchCondition<T> condition;
+        private final SubEvents events;
+
+        ConditionWithEvents(ArchCondition<T> condition, T item) {
+            this(condition, check(condition, item));
+        }
+
+        ConditionWithEvents(ArchCondition<T> condition, SubEvents events) {
+            this.condition = condition;
+            this.events = events;
+        }
+
+        public ConditionEvents getEvents() {
+            return events;
+        }
+
+        private static <T> SubEvents check(ArchCondition<T> condition, T item) {
+            SubEvents events = new SubEvents();
+            condition.check(item, events);
+            return events;
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                    .add("condition", condition)
+                    .add("events", events)
+                    .toString();
+        }
+    }
+
+    abstract static class JoinConditionEvent<T> implements ConditionEvent {
+        final T correspondingObject;
+        final List<ConditionWithEvents<T>> evaluatedConditions;
+
+        JoinConditionEvent(T correspondingObject, List<ConditionWithEvents<T>> evaluatedConditions) {
+            this.correspondingObject = correspondingObject;
+            this.evaluatedConditions = evaluatedConditions;
+        }
+
+        List<String> getUniqueLinesOfViolations() { // TODO: Sort by line number, then lexicographically
+            final Set<String> result = new TreeSet<>();
+            for (ConditionWithEvents<T> evaluation : evaluatedConditions) {
+                for (ConditionEvent event : evaluation.events.getViolating()) {
+                    result.addAll(event.getDescriptionLines());
+                }
+            }
+            return ImmutableList.copyOf(result);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                    .add("evaluatedConditions", evaluatedConditions)
+                    .toString();
+        }
+
+        List<ConditionWithEvents<T>> invert(List<ConditionWithEvents<T>> evaluatedConditions) {
+            return evaluatedConditions.stream().map(this::invert).collect(toList());
+        }
+
+        private ConditionWithEvents<T> invert(ConditionWithEvents<T> evaluation) {
+            SubEvents invertedEvents = new SubEvents();
+            Stream.concat(
+                    evaluation.events.getAllowed().stream(),
+                    evaluation.events.getViolating().stream()
+            ).forEach(event -> invertedEvents.add(event.invert()));
```

Comment:
This could also be written as:
```suggestion
            ).map(ConditionEvent::invert).forEach(invertedEvents::add);            
```
but after a second thought, I'm not even convined that this is better any more, so just pick what you prefer.

Reply:
Not much difference to me either :man_shrugging: So I'll just keep it...